util/src/MultiSizeBufferPool.cpp

```diff
@@ -0,0 +1,688 @@
+#else
+#define DEBUG_PRINT(x, y)
+#endif
+
+#define KVLOG_PREFIX name_, std::this_thread::get_id()
+
+/* ==========================================
+===
+===     Statics / Globals
+===
+========================================== */
+template <typename T>
+static void throwException(const logging::Logger& logger,
+                           const std::string& prefix_message,
+                           const std::string& args_formatted_string) {
+  LOG_ERROR(logger, prefix_message << args_formatted_string);
+  throw T(__PRETTY_FUNCTION__ + std::string(" :") + prefix_message + args_formatted_string);
+}
+
+static inline std::ostream& operator<<(std::ostream& os, const MultiSizeBufferPool::Config& c) {
+  os << " Configuration:"
+     << KVLOG(c.kDefaultBasePeriodicTimerIntervalSec,
+              c.maxAllocatedBytes,
+              c.purgeEvaluationFrequency,
+              c.statusReportFrequency,
+              c.histogramsReportFrequency,
+              c.metricsReportFrequency,
+              c.enableStatusReportChangesOnly,
+              c.subpoolSelectorEvaluationNumRetriesMinThreshold,
+              c.subpoolSelectorEvaluationNumRetriesMaxThreshold);
+  return os;
+}
+
+static inline std::ostream& operator<<(std::ostream& os, const MultiSizeBufferPool::SubpoolConfig& s) {
+  os << " SubpoolConfig:" << KVLOG(s.bufferSize, s.numInitialBuffers, s.numMaxBuffers);
+  return os;
+}
+
+inline std::ostream& operator<<(std::ostream& os, const MultiSizeBufferPool::Statistics& s) {
+  const auto& c = s.current_;
+  const auto& o = s.overall_;
+  os << " Current Statistics: ["
+     << KVLOG(c.numAllocatedBytes_,
+              c.numNonAllocatedBytes_,
+              c.numUnusedBytes_,
+              c.numUsedBytes_,
+              c.numAllocatedChunks_,
+              c.numNonAllocatedChunks_,
+              c.numUsedChunks_,
+              c.numUnusedChunks_)
+     << "]";
+  os << " Overall Statistics: ["
+     << KVLOG(o.numAllocatedBytes_,
+              o.numDeletedBytes_,
+              o.numUsedBytes_,
+              o.numAllocatedChunks_,
+              o.numDeletedChunks_,
+              o.numUsedChunks_,
+              o.numBufferUsageFailed_,
+              o.numBufferUsageSuccess_,
+              o.maxNumAllocatedBytes_,
+              o.maxNumAllocatedChunks_)
+     << "]";
+  return os;
+}
+
+uint32_t MultiSizeBufferPool::getBufferSize(const char* buffer) {
+  char* chunk = bufferToChunk(buffer);
+  auto* chunkHeader = reinterpret_cast<MultiSizeBufferPool::ChunkHeader*>(chunk);
+  return chunkHeader->bufferSize_;
+}
+
+/* ==========================================
+===
+===     MultiSizeBufferPool Definition
+===
+========================================== */
+
+MultiSizeBufferPool::~MultiSizeBufferPool() {
+  timers_.cancel(periodicTimer_);
+  // This one triggers all subpools dtors
+  // Important: we assume here that this is user responsibilty to make sure no more waiting threads -
+  // all buffers have been returned to pool
```

Comment:
Yes it makes sense, i'm doing it in a subpool level. The pool is just a shell. Although it has statistics reported by each subpool, it can't. check this. each subpool is checking in it's dtor if  everything was returned.
Rethinking, writing this is incorrect. We are not sure here they were returned.
This is check in:
`subpools_.clear();`

I'll change this above line to make it clear (we check/warn if all bufffers returned in subpool level)
Thanks.